sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -230,7 +235,7 @@ object PartitioningUtils {
         // Once we get the string, we try to parse it and find the partition column and value.
         val maybeColumn =
           parsePartitionColumn(currentPath.getName, typeInference, userSpecifiedDataTypes,
-            validatePartitionColumns, timeZone)
+            validatePartitionColumns, timeZone, dateFormatter, timestampFormatter)
```

Comment:
Maybe some of you have already asked this question. Do we have any forward compatibility and backward compatibility issue here? For example, can Spark 3.0 return the same value that is wrote by Spark 2.4? and can Spark 2.4 return the same value that is wrote by Spark 3.0?

Reply:
In the frame of Gregorian calendar (> October 1582), Spark 2.4 and Spark 3.0 (new parser) are compatible. For older dates/timestamps, they are not due to difference in average year length: Gregorian calendar is more precise and has 365.2425 days in a year comparing to Julian calendar which has 365.25 days per year.

Reply:
@MaxGekk you probably have a better summary than I do, but is the problem fundamentally about writing formatted dates incorrectly? older dates would have slightly the wrong hour/minute, and was it because the timezone wasn't fully and correctly specified? and this would manifest if writing to JSON or CSV? yeah, a brief summary of the type of bug that was fixed by the new parser would be helpful in the release notes (just "Docs text" of the JIRA)

Reply:
When the partition column is Date type, our parquet writer stores the date value as the directory name. Is that possible that our Spark could return a wrong result? 

For example, 
- Join two partitioned tables using the partitioned key column whose data type is Date. 
- One table is wrote by Spark 2.4 or prior, and another table is wrote by Spark 3.0 or later.
- The partition columns contain the value before October 1582.

Reply:
Let's say we do formatting the same number of days by legacy and new formatter:
- Legacy parser: `-683005` -> `"0100-01-01"`
- New parser: `-683005` -> `"0099-12-30"`

> Is that possible that our Spark could return a wrong result?

The result can be different.

Reply:
I think partition column is fine, as the value is encoded as string, e.g. `0099-12-30`

whatever the formatter we use, the round trip result should be same, e.g. `0099-12-30` -> internal representation(legacy and new formatter can be different) -> `0099-12-30`

Reply:
If the data is read and wrote by the same version, it is fine. The problem is the data could be written and read by different version of Spark. The case I shown above is to join two tables that are written by different versions.

Reply:
The value of partition column is spark version independent. When Spark sees `0099-12-30`, it doesn't care who wrote it, and will covert `0099-12-30` into number of days.

Reply:
Since the output values are different, the results are still different from the expected, right?

Reply:
I would guess the result would be the same till only textual representation of dates is involved into a query. Let's say if you read dates from a datasource:
```
date,id
2019-01-03,1
1019-01-03,2
1019-01-03,3
0019-01-03,4
```
```scala
scala> val df1 = spark
  .read.option("header", true).option("inferSchema", true).csv("dates")
  .where('date < "1582-10-01")
  .groupBy('date)
  .agg(max('id))
df1: org.apache.spark.sql.DataFrame = [date: string, max(id): int]

scala> df1.show
+----------+-------+
|      date|max(id)|
+----------+-------+
|0019-01-03|      4|
|1019-01-03|      3|
+----------+-------+
```
```scala
scala> val df2 = Seq(("0119-01-03", 6), ("1019-01-03", 7)).toDF("date", "id")
df2: org.apache.spark.sql.DataFrame = [date: string, id: int]

scala> df2.show
+----------+---+
|      date| id|
+----------+---+
|0119-01-03|  6|
|1019-01-03|  7|
+----------+---+
```
```scala
scala> val joined = df1.join(df2, df1("date") === df2("date")).select(df1("date"), 'id, $"max(id)"); joined.show
+----------+---+-------+
|      date| id|max(id)|
+----------+---+-------+
|1019-01-03|  7|      3|
+----------+---+-------+
```
```scala
joined.write.partitionBy("date").csv("date2")
```
```shell
$ tree date2
.
├── _SUCCESS
└── date=1019-01-03
    ├── _SUCCESS
    ├── _committed_2042414757166556120
    ├── _started_2042414757166556120
    └── part-00001-tid-2042414757166556120-5729e821-9cf3-41f1-94af-c909bd0f9972-1621.c000.csv

1 directory, 5 files
```
The same result on Spark 2.4 and this branch.